glean-core/src/lib.rs

```diff
@@ -195,21 +193,35 @@ impl Glean {
             max_events: cfg.max_events.unwrap_or(DEFAULT_MAX_EVENTS),
             is_first_run: false,
         };
-        glean.on_change_upload_enabled(cfg.upload_enabled);
+
+        // Set the initial state of upload enabled. If we don't have any client
+        // id, assume this is a new installation and set to the default of True.
+        // If we have a client_id, if it's the KNOWN_CLIENT_ID, we know that
+        // upload was disabled previously. This is changed to the value
+        // requested by the user from the language binding immediately after
+        // initialization. This allows for the detection of changes to this
+        // preference outside of the run of the applicaton, such as in a
+        // configuration file and allows changes such as clearing the database
+        // and sending a deletion ping to happen. See bug 1624695.
+        glean.upload_enabled = match glean
+            .core_metrics
+            .client_id
+            .get_value(&glean, "glean_client_info")
+        {
+            None => true,
```

Comment:
What if we keep all the existing code and change this part slightly: let's only find if we have an existing non coffee client id. If so, forcefully trigger the deletion ping here, as done in [there](https://github.com/mozilla/glean/blob/2b5ea921f814c4f145793bbed061b58cc573685e/glean-core/src/lib.rs#L302-L306). This would require us to deteramine is_first_run a bit earlier.

Moreover, wouldn't the current approach generate a new deletion ping the first time a new profile starts the product with telemetry disabled?

Reply:
> What if we keep all the existing code and change this part slightly: let's only find if we have an existing non coffee client id. If so, forcefully trigger the deletion ping here, as done in there. This would require us to deteramine is_first_run a bit earlier.

I tried that initially, but the problem is that we need to *both* submit the deletion ping (which happens in Rust) *and* trigger the deletion ping uploader (from the language bindings).

I have some thoughts about refactoring this further to make the line between the Rust and language binding side less weird here.

> Moreover, wouldn't the current approach generate a new deletion ping the first time a new profile starts the product with telemetry disabled?

Yes, that's true.  Should work on that case as well, I guess.